File: src/connector/odbc/tests/odbc.py
import pyodbc
# cnxn = pyodbc.connect('DSN={TAOS_DSN};UID={ root };PWD={ taosdata };HOST={ localhost:6030 }', autocommit=True)
cnxn = pyodbc.connect('DSN={TAOS_DSN}; UID=root;PWD=taosdata; HOST=localhost:6030', autocommit=True)
cnxn.setdecoding(pyodbc.SQL_CHAR, encoding='utf-8')
#cnxn.setdecoding(pyodbc.SQL_WCHAR, encoding='utf-8')
#cnxn.setencoding(encoding='utf-8')

#cursor = cnxn.cursor()
#cursor.execute("SELECT * from db.t")
#row = cursor.fetchone()
#while row:
#    print(row)
#    row = cursor.fetchone()
#cursor.close()

#cursor = cnxn.cursor()
#cursor.execute("""
#INSERT INTO db.t values (?,?,?,?,?,?,?,?,?,?)
#""",
#"2020-12-12 00:00:00",
#1,
#27,
#32767,
#147483647,
#223372036854775807,
#23.456,
#899.999999,
#"foo",
#"bar")

cursor = cnxn.cursor()
cursor.execute("drop database if exists db");
cursor.close()

cursor = cnxn.cursor()
cursor.execute("create database db");
cursor.close()

cursor = cnxn.cursor()
cursor.execute("create table db.mt (ts timestamp, b bool, v1 tinyint, v2 smallint, v4 int, v8 bigint, f4 float, f8 double, bin binary(40), blob nchar(10))");
cursor.close()

cursor = cnxn.cursor()
cursor.execute("insert into db.mt values('2020-10-13 06:44:00', 1, 127, 32767, 32768, 32769, 123.456, 789.987, 'hello', 'world')")
cursor.close()

cursor = cnxn.cursor()
cursor.execute("insert into db.mt values(?,?,?,?,?,?,?,?,?,?)", "2020-10-13 07:06:00", 0, 127, 32767, 32768, 32769, 123.456, 789.987, "hel后lo", "wo哈rld");
cursor.close()

cursor = cnxn.cursor()
cursor.execute("SELECT * from db.mt")
row = cursor.fetchone()
while row:
    print(row)
    row = cursor.fetchone()
cursor.close()

#cursor = cnxn.cursor()
#cursor.execute("drop database if exists db");
#cursor.close()
#
#cursor = cnxn.cursor()
#cursor.execute("create database db");
#cursor.close()

cursor = cnxn.cursor()
cursor.execute("create table db.t (ts timestamp, b bool, v1 tinyint, v2 smallint, v4 int, v8 bigint, f4 float, f8 double, bin binary(4), blob nchar(4))");
cursor.close()

cursor = cnxn.cursor()
cursor.execute("insert into db.t values('2020-10-13 06:44:00', 1, 127, 32767, 32768, 32769, 123.456, 789.987, 'hell', 'worl')")
cursor.close()

cursor = cnxn.cursor()
cursor.execute("SELECT * from db.t")
row = cursor.fetchone()
while row:
    print(row)
    row = cursor.fetchone()
cursor.close()

cursor = cnxn.cursor()
cursor.execute("create table db.v (ts timestamp, v1 tinyint)")
cursor.close()

params = [ ('A', 1), ('B', 2), ('C', 3) ]
params = [ ('A', 1), ('B', 2), ('C', 3) ]
params = [ ('2020-10-16 00:00:00', 1),
           ('2020-10-16 00:00:01', 4),
           ('2020-10-16 00:00:02', 5),
           ('2020-10-16 00:00:03.009', 6) ]
cursor = cnxn.cursor()
cursor.fast_executemany = True
cursor.executemany("insert into db.v values (?, ?)", params)
cursor.close()

cursor = cnxn.cursor()
cursor.execute("SELECT * from db.v")
row = cursor.fetchone()
while row:
    print(row)
    row = cursor.fetchone()
cursor.close()

cursor = cnxn.cursor()
cursor.execute("SELECT * from db.v where v1 > ?", 4)
row = cursor.fetchone()
while row:
    print(row)
    row = cursor.fetchone()
cursor.close()

cursor = cnxn.cursor()
cursor.execute("SELECT * from db.v where v1 > ?", '5')
row = cursor.fetchone()
while row:
    print(row)
    row = cursor.fetchone()
cursor.close()

cursor = cnxn.cursor()
cursor.execute("create table db.f (ts timestamp, v1 float)")
cursor.close()

params = [ ('2020-10-20 00:00:10', '123.3') ]
cursor = cnxn.cursor()
cursor.fast_executemany = True
cursor.executemany("insert into db.f values (?, ?)", params)
cursor.close()

File: 2.0/src/connector/odbc/examples/js/odbc.js
#!/usr/bin/env node

const odbc = require('odbc');
const path = require('path');

function usage() {
  var arg = path.basename(process.argv[1]);
  console.error(`usage:`);
  console.error(`${arg} --DSN <DSN> --UID <uid> --PWD <pwd> --Server <host:port>`);
  console.error(`${arg} -C <conn_str>`);
  console.error(`  conn_str eg: 'DSN={TAOS_DSN};UID=root;PWD=taosdata;Server=host:port'`);
}

var cfg = { };

if (process.argv.length==2) {
    usage();
    process.exit(0);
}

var i;
for (i=2; i<process.argv.length; ++i) {
  var arg = process.argv[i];
  if (arg=='-h') {
    usage();
    process.exit(0);
  }
  if (arg=="--DSN") {
    ++i;
    if (i>=process.argv.length) {
      console.error(`expecting <dns> after --DSN but got nothing`);
      usage(process.argv[1]);
      process.exit(1);
    }
    arg = process.argv[i];
    cfg.dsn = arg;
    continue;
  }
  if (arg=="--UID") {
    ++i;
    if (i>=process.argv.length) {
      console.error(`expecting <uid> after --UID but got nothing`);
      usage(process.argv[1]);
      process.exit(1);
    }
    arg = process.argv[i];
    cfg.uid = arg;
    continue;
  }
  if (arg=="--PWD") {
    ++i;
    if (i>=process.argv.length) {
      console.error(`expecting <pwd> after --PWD but got nothing`);
      usage(process.argv[1]);
      process.exit(1);
    }
    arg = process.argv[i];
    cfg.pwd = arg;
    continue;
  }
  if (arg=="--Server") {
    ++i;
    if (i>=process.argv.length) {
      console.error(`expecting <host:port> after --Server but got nothing`);
      usage(process.argv[1]);
      process.exit(1);
    }
    arg = process.argv[i];
    cfg.server = arg;
    continue;
  }
  if (arg=="-C") {
    ++i;
    if (i>=process.argv.length) {
      console.error(`expecting <conn_str> after -C but got nothing`);
      console.error(`  conn_str eg: 'DSN={TAOS_DSN};UID=root;PWD=taosdata;Server=host:port'`);
      process.exit(1);
    }
    arg = process.argv[i];
    cfg.conn_str = arg;
    continue;
  }
  console.error(`unknown argument: [${arg}]`);
  usage(process.argv[1]);
  process.exit(1);
}

var connectionString = cfg.conn_str;

if (!cfg.conn_str) {
  connectionString = `DSN={${cfg.dsn}}; UID=${cfg.uid}; PWD=${cfg.pwd}; Server=${cfg.server}`;
}

(async function () {
    const connStr = connectionString;
    try {
      console.log(`connecting [${connStr}]...`);
      const connection = await odbc.connect(connStr);
      await connection.query('create database if not exists m');
      await connection.query('use m');
      await connection.query('drop table if exists t');
      await connection.query('create table t (ts timestamp, b bool, v1 tinyint, v2 smallint, v4 int, v8 bigint, f4 float, f8 double, bin binary(10), name nchar(3))');
      await connection.query('insert into t values("2020-01-02 12:34:56.781", 1, 127, 32767, 32768, 32769, 123.456, 789.987, "hello", "我和你")');
      console.log('.........');
      result = await connection.query('select * from t');
      console.log(result[0]);


      statement = await connection.createStatement();
      await statement.prepare('INSERT INTO t (ts, v1) VALUES(?, ?)');
      await statement.bind(['2020-02-02 11:22:33.449', 89]);
      result = await statement.execute();
      console.log(result);

      result = await connection.query('select * from t');
      console.log(result[0]);
      console.log(result[1]);
    } catch (e) {
      console.log('error:', e);
    }
})();

